### DISCUSSION TITLE
Upgrade (or replace) codemirror uses

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#24898] [Feature Requests]
CodeMirror 6 has launched, which is a pretty substantial upgrade from CodeMirror 5. That being said, it's not a drop-in replacement, or an easy migration path. We should take the time to visit all our uses of codemirror and upgrade to CodeMirror 6.  ...or upgrade to something else entirely. In the meantime, [the Monaco editor](https://microsoft.github.io/monaco-editor/index.html) has become way more stable and feature rich as well, which could be a good alternative.

**Comment by josephdpurcell** [comment]
Here is the migration guide from CodeMirror5 to 6: https://codemirror.net/docs/migration/.

Here's what I tried:

```
git clone this repo
cd directus/api
pnpm uninstall codemirror @types/codemirror
pnpm i -D codemirror@6
pnpm --filter @directus/app dev
```

And it prints out a ton of errors, which is expected. It looks like all of them are like `Missing X export in "codemirror" package` like so:

```
✘ [ERROR] [plugin vite:dep-scan] Missing "./mode/meta" export in "codemirror" package

    src/interfaces/input-code/index.ts:4:7:
      4 │ import 'codemirror/mode/meta';
        ╵        ~~~~~~~~~~~~~~~~~~~~~~
```

Here are the most common references to the pattern `import.*codemirr`:

```
    122 ./app/src/interfaces/input-code/import-codemirror-mode.ts
     23 ./app/src/interfaces/input-code/input-code.vue
      3 ./app/src/interfaces/input-rich-text-md/input-rich-text-md.vue
      2 ./app/src/interfaces/_system/system-raw-editor/system-raw-editor.vue
      2 ./app/src/interfaces/input-code/index.ts
      1 ./app/src/styles/main.scss
      1 ./app/src/interfaces/input-rich-text-md/edits.ts
```

Some things that stick out to me, and I speak here as *very* novice with Directus:

* The main.scss imports `lib/codemirror`. Looking at `./app/src/styles/lib/_codemirror.scss` theres about 500 lines of non-comment code. Given CodeMirror6 is a complete rewrite it's probably safest to assume this effort would need re-done.
* The Code field (`input-code`) has a deep integration with CodeMirror since it allows selecting various languages. The language system was re-done, so it may be safer to start from scratch with this field?
* The Markdown field (`input-rich-text-md`) is hopefully a light refactor. Looking at the Directus field now compared to what I did to setup a basic CM6 field in this comment https://github.com/codemirror/dev/issues/1020#issuecomment-1343082372 it seems like it would be straightforward. But, details matter.
* The `system-raw-editor` I'm not sure about. I see references to it in `app/src/components/v-form/form-field-raw-editor.vue` and `app/src/components/v-form/form-field-interface.vue`.

My gut reaction to this is the upgrade path is substantial, but doable though not by a Directus novice. At least on the surface it seems like CodeMirror6 would support the desired functionality of Directus' uses for CM.

Regarding the Markdown field specifically, I haven't been fully satisfied with Directus' Markdown field experience (notably https://github.com/directus/directus/issues/16690), and have started a discussion about what the "best" Markdown field experience might be over here: https://github.com/directus/directus/discussions/16762. But, the "best" Markdown authoring experience may not overlap with the "best" code writing experience, as I noted https://github.com/directus/directus/issues/16690#issuecomment-1335918841, because Markdown is for the non-technical (see also RFC). Thus, it may be best to have a separate discussion just for the Code field.

**Comment by rijkvanzanten** [comment]
Linear: [ENG-160](https://linear.app/directus/issue/ENG-160/upgrade-or-replace-codemirror-uses)

**Comment by josephdpurcell** [comment]
Side Note: Agit CMS is using CodeMirror6 and the authoring experience is really nice. This suggests to me that moving forward with an upgrade to CodeMirror6 may result in an improved authoring experience for Directus as well. See related Agit CMS ticket: https://github.com/0xsuk/agitcms/issues/30#issuecomment-1366030273

**Comment by rijkvanzanten** [comment]
> Side Note: Agit CMS is using CodeMirror6 and the authoring experience is really nice. This suggests to me that moving forward with an upgrade to CodeMirror6 may result in an improved authoring experience for Directus as well. See related Agit CMS ticket: [0xsuk/agitcms#30 (comment)](https://github.com/0xsuk/agitcms/issues/30#issuecomment-1366030273)

Sweet!

Also familiar face and title on that linked issue ;P
